### PR TITLE
cross-platform aync multi-client solution

### PR DESCRIPTION
**Pull request by bubbleboy14**
I modified WebSocketApp.run_forever(), replacing the select() loop with rel.read() event registration. Info on rel here:

https://github.com/bubbleboy14/registeredeventlistener

This addresses platform issues (e.g. https://github.com/websocket-client/websocket-client/pull/293/files) by choosing the fastest async method (such as epoll, poll, kqueue, select) supported by the system.

Furthermore, one can now run more than one WebSocketApp by simply passing no_dispatch=True to run_forever() as many times as are necessary (n-1 times -> all but the last, which will finally start the event loop).

### CODE REVIEW

**Comment by liris** [comment]
rel doesn’t look maintained, anymore.
Do you  have another option?

**Comment by bubbleboy14** [comment]
I still maintain it.

**Comment by bubbleboy14** [comment]
Try firing off multiple clients, it's fun, like:

"""
def ws_app_creator(chan):
    ws = websocket.WebSocketApp(chan, on_message=on_message)
    ws.run_forever(no_dispatch=True)

map(ws_app_creator, channels_array)
rel.dispatch()
"""

**Comment by liris** [comment]
Where is the project page?

**Comment by liris** [comment]
Sorry, I misunderstood.

Looks nice.

**Comment by liris** [comment]
Do you have test code for this?

**Comment by bubbleboy14** [comment]
simple multi-client example:

```
import rel, websocket

def on_message(ws, message):
	print ws.symbol, message

def spawn(symbol):
	ws = websocket.WebSocketApp("wss://api.gemini.com/v1/marketdata/%s"%(symbol,),
		on_message=on_message)
	ws.symbol = symbol
	ws.run_forever(no_dispatch=True)

map(spawn, ["BTCUSD", "ETHUSD"])
rel.dispatch()
```

**Comment by invokr** [comment]
Will this get merged? I'm also running into the 1024 FD max limit for a long running process that spawns connections over a long time.

**Comment by bubbleboy14** [comment]
@invokr by Jan 23: https://github.com/websocket-client/websocket-client/milestone/2

**Comment by invokr** [comment]
thanks @bubbleboy14, missed the milestone.

**Comment by liris** [comment]
works fine on 2.7.14

**Comment by liris** [comment]
@bubbleboy14 I tried to install rel module to Python3.6.4, but I got errors.

```shell
Collecting rel
  Using cached rel-0.3.1.tar.gz
    Complete output from command python setup.py egg_info:
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/private/var/folders/q5/hrw3j3kj13zdg5r9xs8r_ty4wp33hf/T/pip-build-e7t47o55/rel/setup.py", line 1, in <module>
        from rel import __version__
      File "/private/var/folders/q5/hrw3j3kj13zdg5r9xs8r_ty4wp33hf/T/pip-build-e7t47o55/rel/rel/__init__.py", line 2, in <module>
        from rel import override, supported_methods, initialize, read, write, timeout, signal, event, dispatch, loop, abort, init, sys, EV_PERSIST, EV_READ, EV_SIGNAL, EV_TIMEOUT, EV_WRITE
    ImportError: cannot import name 'override'
```

Does it support python3 or something missing?

**Comment by bubbleboy14** [comment]
I see you have an older version (0.3.1) -- does it work with the current version, 0.3.6?

**Comment by bubbleboy14** [comment]
@liris ah, the 0.3 series wasn't p3k compliant. I just pushed a new version, 0.4.0, that is. Works fine with my tests. Does it work for you? Thanks :)

**Comment by liris** [comment]
I tried the following code. Python 2.7 worked fine, but python 3.6 doesn't.
The program terminated immediately without blocking at all. Could you check it or something wrong with me?

```
import websocket
import rel
import threading
import time

no_dispatch=True

def on_message(ws, message):
    print(ws.symbol, message)

def runIt(ws):
    time.sleep(1)
    ws.send("iayn")

def spawn(symbol):
    ws = websocket.WebSocketApp("ws://echo.websocket.org",
		on_message=on_message)
    ws.symbol = symbol
    threading.Thread(target=runIt, args=(ws,)).start()
    ws.run_forever(no_dispatch=no_dispatch)

map(spawn, ["BTCUSD", "ETHUSD"])
rel.dispatch()
```

**Comment by bubbleboy14** [comment]
@liris I should have mentioned, the second-to-last line there, """map(spawn, ["BTCUSD", "ETHUSD"])""", should be changed to """list(map(spawn, ["BTCUSD", "ETHUSD"]))""" for it to work in py3

**Comment by liris** [comment]
Thanks, it works on py3

File: websocket/_app.py [hunk 1]
```diff
@@ -141,7 +142,7 @@ def run_forever(self, sockopt=None, sslopt=None,
                     http_proxy_host=None, http_proxy_port=None,
                     http_no_proxy=None, http_proxy_auth=None,
                     skip_utf8_validation=False,
-                    host=None, origin=None):
+                    host=None, origin=None, no_dispatch=False):
```
Comment (liris) [inline websocket/_app.py]: Could you provide document for no_dispatch parameter?

**Comment by bubbleboy14** [comment]
"Pass 'no_dispatch=True' to omit event loop dispatch; this allows further code execution, including initialization of additional websocket clients"

**Comment by liris** [comment]
I changed the interface.
Please use like this.

```python
import websocket
import rel
import threading
import time

no_dispatch=True

class RelDispatcher:
    def read(self, sock, callback):
        rel.read(sock, callback)
        rel.signal(2, rel.abort)

    def dispatch(self):
        # rel.dispatch()
        pass

def on_message(ws, message):
    print(ws.symbol, message)

def runIt(ws):
    time.sleep(1)
    ws.send("iayn")

def spawn(symbol):
    ws = websocket.WebSocketApp("ws://echo.websocket.org",
		on_message=on_message)
    ws.symbol = symbol
    threading.Thread(target=runIt, args=(ws,)).start()
    ws.run_forever(dispatcher=RelDispatcher())
    print("####")

list(map(spawn, ["BTCUSD", "ETHUSD"]))
rel.dispatch()

```

**Comment by bubbleboy14** [comment]
Ooo, good call @liris, that's a lot cleaner than my no_dispatch idea. I have one more suggestion https://github.com/websocket-client/websocket-client/pull/389) for making this even better. What do you think?

**Pull request merged**